src/atomic/atomic_cell.rs

```diff
@@ -0,0 +1,695 @@
+use core::cell::UnsafeCell;
+use core::fmt;
+use core::mem;
+use core::ptr;
+use core::slice;
+use core::sync::atomic::{self, AtomicBool, Ordering};
+
+/// A thread-safe mutable memory location.
+///
+/// This type is equivalent to [`Cell`], except it can also be shared among multiple threads.
+///
+/// Operations on `AtomicCell`s use atomic instructions whenever possible, and synchronize using
+/// global locks otherwise. You can call [`AtomicCell::<T>::is_lock_free()`] to check whether
+/// atomic instructions or locks will be used.
+///
+/// [`Cell`]: https://doc.rust-lang.org/std/cell/struct.Cell.html
+/// [`AtomicCell::<T>::is_lock_free()`]: struct.AtomicCell.html#method.is_lock_free
+pub struct AtomicCell<T> {
+    /// The inner value.
+    ///
+    /// If this value can be transmuted into a primitive atomic type, it will be treated as such.
+    /// Otherwise, all potentially concurrent operations on this data will be protected by a global
+    /// lock.
+    value: UnsafeCell<T>,
```

Comment:
Ideally we would use alignment to round this up to the size of the next atomic type. This is what `_Atomic` in C11 does. However there is currently no way to express this in Rust.

Reply:
@Amanieu by "next atomic type" what do you mean?

Reply:
Probably that alignment/size of `_Atomic char[3]` is the same as alignment/size of `_Atomic uint64_t` (because that's how it gets represented internally).